### PR TITLE
[NeMo Megatron] use worker processes for data preprocessing

### PR DESCRIPTION
**Pull request by crcrpar**
# What does this PR do ?

Update Megatron style `BatchSampler`s and `GPTDataset` to prepare data samples in worker processes.

# Changelog 
- BatchSamplers return a list of `num_micro_batch x micro_batch_size` indices: https://github.com/NVIDIA/NeMo/compare/main...crcrpar:megatron/update_data_loading?expand=1#diff-a78cf82731c4e37f487037ab45cab0183aac6216ee96e08d3b81ff111495d3caR36 and https://github.com/NVIDIA/NeMo/compare/main...crcrpar:megatron/update_data_loading?expand=1#diff-a78cf82731c4e37f487037ab45cab0183aac6216ee96e08d3b81ff111495d3caR106
- A new arguments of `eos_id` is added to `GPTDataset` in order to change the callsite of `get_ltor_masks_and_position_ids` to `GPTDataset`. [`eos_id`](url) is an attribute of a `Tokenizer` instance.
- `process_micro_batch` is removed in favor of updated `GPTDataset.__getitem__`: https://github.com/NVIDIA/NeMo/compare/main...crcrpar:megatron/update_data_loading?expand=1#diff-e93ccae74f4b67d341676afc9f3c7e2c50f751ec64df84eb3b2a86b62029ef76L456
- `NLPDataConnector` isn't needed thanks to the updated `GPTDataset` and `BatchSampler`s: https://github.com/NVIDIA/NeMo/compare/main...crcrpar:megatron/update_data_loading?expand=1#diff-424bb19c0b7c9dba2aebb3459972c5a8bf8c902dca48a8762f12bc5cde5fea5fL68

# Usage

To utilize worker processes, it'd be better to set `model.data.num_workers` > 0.
See https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by crcrpar** [comment]
okay now this breaks T5 dataset, will fix https://lgtm.com/projects/g/NVIDIA/NeMo/snapshot/dbf66965fe911aa3969f64d16c3ca7c70fc54f44/files/nemo/collections/nlp/data/language_modeling/megatron/lm_adapted_t5_dataset.py?sort=name&dir=ASC&mode=heatmap#xd18df39f4b694888:1

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py [hunk 1]
```diff
@@ -162,7 +183,19 @@ def get_indexed_dataset_(data_prefix, data_impl, skip_warmup):
 
 
 class GPTDataset(MegatronDataset):
-    def __init__(self, cfg, trainer, name, data_prefix, documents, indexed_dataset, num_samples, seq_length, seed):
+    def __init__(
```
Comment (crcrpar) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py:186]: @MaximumEntropy Could you look at this updated `GPTDataset` as T5 dataset inherits?

File: examples/nlp/language_modeling/megatron_gpt_pretraining.py [hunk 2]
```diff
@@ -63,10 +62,6 @@ def main(cfg) -> None:
 
     trainer = Trainer(plugins=plugins, **cfg.trainer)
 
-    # NLPDataConnector used to provide global batches which are needed
-    # for Apex fwd/bwd functions
-    trainer._data_connector = NLPDataConnector(trainer)
```
Comment (ericharper) [inline examples/nlp/language_modeling/megatron_gpt_pretraining.py:68]: Can you totally remove `NLPDataConnector` ? I mean remove it from nlp_overrides.py as well.

Comment (crcrpar) [reply examples/nlp/language_modeling/megatron_gpt_pretraining.py:68]: done in https://github.com/NVIDIA/NeMo/pull/3665/commits/94e8d0c48a0db199500534ea47ae2e62518a3399

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py (hunk 1 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py:186]: @crcrpar I can create a PR to this branch that fixes this for T5LMAdaptedDataset.

Comment (crcrpar) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py:186]: Does that mean the change for T5LMAdaptedDataset in this PR isn't enough?

File: nemo/collections/nlp/data/language_modeling/megatron/_megatron_batch_samplers.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        self._num_micro_batch_times_micro_batch_size = num_micro_batch_times_micro_batch_size
+        self.data_parallel_rank = data_parallel_rank
+        self.data_parallel_size = data_parallel_size
+        self.num_micro_batch_times_micro_batch_size_times_data_parallel_size = (
+            self._num_micro_batch_times_micro_batch_size * data_parallel_size
+        )
+        self.drop_last = drop_last
+
+    def __len__(self):
+        return self.total_samples
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/data/language_modeling/megatron/_megatron_batch_samplers.py:74]: Why is the `__len__` different from before?

Comment (MaximumEntropy) [reply nemo/collections/nlp/data/language_modeling/megatron/_megatron_batch_samplers.py:74]: Earlier it was `return (self.total_samples - self.consumed_samples - 1) // self.micro_batch_times_data_parallel_size + 1`.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py [hunk 4]
```diff
@@ -208,8 +237,59 @@ def __getitem__(self, idx):
...
+            'loss_mask': loss_mask,
+            'position_ids': position_ids,
+        }
+
+
+@torch.no_grad()
+def _create_ltor_masks_and_position_ids(
+    tokens: torch.Tensor, eod_token: int, reset_position_ids: bool, reset_attention_mask: bool, eod_mask_loss: bool,
+):
+    """Modified version of get_ltor_masks_and_position_ids in nemo/collections/nlp/modules/common/megatron/utils.py"""  # NOQA
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py:263]: Can you make the comment a little more clear on how/why it was modified for micro vs global batch.

**Review by MaximumEntropy — COMMENTED** [review]
Looks good! Just a couple of minor comments.

File: nemo/collections/nlp/data/language_modeling/megatron/_megatron_batch_samplers.py (hunk 3 above)
Comment (crcrpar) [reply nemo/collections/nlp/data/language_modeling/megatron/_megatron_batch_samplers.py:74]: nice catch. I think I just wrote this to meet the BatchSampler's requirements.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py (hunk 1 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py:186]: No, it's good. I didn't realize you already made the change by the time I commented. I was still looking at the LGTM link you shared.

File: nemo/collections/nlp/data/language_modeling/megatron/_megatron_batch_samplers.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
+import abc
```
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/_megatron_batch_samplers.py:1]: Can you please copy paste the copyright from another file in nemo to this file?

Comment (ericharper) [reply nemo/collections/nlp/data/language_modeling/megatron/_megatron_batch_samplers.py:1]: Also, can you rename this file to `megatron_batch_samplers.py` We don't have filenames starting with `_` in nemo.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**